pytorch_lightning/utilities/dataloader_fetcher.py

```diff
@@ -0,0 +1,122 @@
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Any, Callable, Generator, List
+
+import torch
+
+import pytorch_lightning as pl
+
+
+@dataclass()
+class LightningStreamEvent:
+
+    """
+    This class is used to capture the device stream and report its associated event
+    """
+
+    inter_batch_parallelism: bool
+    device: torch.device
+
+    def __post_init__(self):
+        if self.cuda_inter_batch_parallelism:
+            self.cuda_stream = torch.cuda.Stream()
+            self.events: List[torch.cuda.Event] = []
+
+    @contextmanager
+    def stream_context(self):
+        if self.cuda_inter_batch_parallelism:
+            with torch.cuda.stream(self.cuda_stream):
+                self.start_record()
+                yield
+                self.end_record()
+        else:
+            self.start_record()
+            yield
+            self.end_record()
+
+    @property
+    def cuda_inter_batch_parallelism(self) -> bool:
+        return self.device.type == "cuda" and self.inter_batch_parallelism
```

Comment:
Would be better to crash when `inter_batch_parallelism` for non GPU Accelerators.

Reply:
Yes, good idea.